mypy/checkexpr.py

```diff
@@ -2674,6 +2674,12 @@ def overload_arg_similarity(actual: Type, formal: Type) -> int:
             return overload_arg_similarity(actual.ret_type, formal.item)
         else:
             return 0
+    if isinstance(actual, TypedDictType):
+        if isinstance(formal, TypedDictType):
+            # Don't support overloading based on the keys or value types of a TypedDict since
+            # that would be complicated and probably only marginally useful.
```

Comment:
Your comment below is correct -- signatures are considered to be overlapping. I created a new issue: #3618.

Note that there are other similar issues such as not being able to overload based on tuple structure, I think. I don't remember hearing user complaints about them, so this is perhaps not a high-priority issue. Anyway let's wait for user feedback before we do more work on this.